lib/datasource.js

```diff
@@ -471,60 +471,46 @@ DataSource.prototype.defineScopes = function(modelClass, scopes) {
 DataSource.prototype.defineRelations = function(modelClass, relations) {
   var self = this;
 
-  // Create a function for the closure in the loop
-  var createListener = function(name, relation, targetModel, throughModel) {
+  // Wait for target/through models to be attached before setting up the relation
+  var deferRelationSetup = function(relationName, relation, targetModel, throughModel) {
     if (!isModelDataSourceAttached(targetModel)) {
-      targetModel.once('dataAccessConfigured', function(model) {
+      targetModel.once('dataAccessConfigured', function(targetModel) {
         // Check if the through model doesn't exist or resolved
         if (!throughModel || isModelDataSourceAttached(throughModel)) {
           // The target model is resolved
           var params = traverse(relation).clone();
-          params.as = name;
-          params.model = model;
+          params.as = relationName;
+          params.model = targetModel;
           if (throughModel) {
             params.through = throughModel;
           }
-          modelClass[relation.type].call(modelClass, name, params);
+          modelClass[relation.type].call(modelClass, relationName, params);
         }
       });
     }
 
     if (throughModel && !isModelDataSourceAttached(throughModel)) {
       // Set up a listener to the through model
-      throughModel.once('dataAccessConfigured', function(model) {
+      throughModel.once('dataAccessConfigured', function(throughModel) {
         if (isModelDataSourceAttached(targetModel)) {
           // The target model is resolved
           var params = traverse(relation).clone();
-          params.as = name;
+          params.as = relationName;
```

Comment:
here the `as` keyword of `params.as` is on root params, not on `params.polymorphic`, this should not be changed.

there is no more reference to `polymorphic.as` in this part of the code.

as for the name, i see your concern. Should we go with `ref`? do you have a suggestion?
I first started with `prefix` but it does not feel quite correct when used in polymorphic through relations